File: Subsets.java
class Solution {
    List<List<Integer>> res;
    public List<List<Integer>> subsets(int[] nums) {
        this.res = new ArrayList<>();
        helper(nums, 0, new ArrayList<>());
        return res;


        
    }
    private void helper(int[] nums, int pivot, List<Integer> path){
       
            res.add(new ArrayList<>(path));
       
        
       
      
       for(int i=pivot;i<nums.length;i++){
         //choose
        path.add(nums[i]);
        helper(nums, i+1, (path));

        path.remove(path.size()-1);
       }
        
    }
}

File: Subsets.py
# TC: O(2^N)
# SC: O(N)

class Solution:
    res=[]

    def helper(self,nums,pivot,li):
        if pivot==len(nums)+1:
            return

        self.res.append(li.copy())

        for i in range(pivot,len(nums)):
            # lis=li.copy()
            li.append(nums[i])
            self.helper(nums,i+1,li)
            li.pop()

        
    def subsets(self, nums: List[int]) -> List[List[int]]:
        self.res=[]
        self.helper(nums,0,[])
        return self.res

        